diff-models/core/src/model.ts

```diff
@@ -0,0 +1,635 @@
+        };
+        this.validatorQ.push(unval);
+        this.m.ccvP.afterUnbondingInitiated(this.opID);
+        this.opID += 1;
+        this.status[i] = Status.UNBONDING;
+      });
+
+    // Compute updates
+    this.changes = {};
+    newVals.forEach((i) => {
+      if (this.tokens[i] != this.lastTokens[i]) {
+        // validator power changed
+        this.changes[i] = this.tokens[i];
+      }
+    });
+    _.difference(newVals, oldVals)
+      .sort((a, b) => a - b)
+      .forEach((i) => {
+        // validator bonded
+        this.changes[i] = this.tokens[i];
+      });
+    _.difference(oldVals, newVals)
+      .sort((a, b) => a - b)
+      .forEach((i) => {
+        // validator no longer bonded
+        this.changes[i] = 0;
+      });
+
+    // Save the valset and their tokens
+    // (mimics block commit)
+    this.lastVals = newVals;
+    this.lastTokens = _.clone(this.tokens);
+  };
+
+  endBlockMaturation = () => {
+    // Process unbonding validators
+    const completedUnvals = this.validatorQ.filter(
+      (e: Unval) =>
+        e.unbondingTime <= this.m.t[P] &&
+        e.unbondingHeight <= this.m.h[P] &&
+        !e.onHold,
+    );
+    completedUnvals.forEach((e: Unval) => {
+      this.status[e.val] = Status.UNBONDED;
+      this.m.events.push(Event.COMPLETE_UNVAL_IN_ENDBLOCK);
+    });
+    this.validatorQ = this.validatorQ.filter(
+      (e) => !completedUnvals.includes(e),
+    );
+
+    // Process undelegations
+    const expiredUndels = this.undelegationQ.filter(
+      (e) => e.completionTime <= this.m.t[P] && !e.expired,
+    );
+    expiredUndels.forEach((e: Undelegation) => (e.expired = true));
+    const completedUndels = expiredUndels.filter((e) => !e.onHold);
+    if (completedUndels.length < expiredUndels.length) {
+      this.m.events.push(Event.SOME_UNDELS_EXPIRED_BUT_NOT_COMPLETED);
+    }
+    this.undelegationQ = this.undelegationQ.filter(
+      (e: Undelegation) => !completedUndels.includes(e),
+    );
+    if (0 < completedUndels.length) {
+      this.m.events.push(Event.COMPLETE_UNDEL_IN_ENDBLOCK);
+    }
+    // Refund completed undelegations
+    this.delegatorTokens += completedUndels.reduce(
+      (x, e) => x + e.balance,
+      0,
+    );
+  };
+
+  endBlock = () => {
+    this.endBlockComputeValUpdates();
+    this.endBlockMaturation();
+  };
+
+  delegate = (val: Validator, amt: number) => {
+    this.delegatorTokens -= amt;
+    this.tokens[val] += amt;
+    this.delegation[val] += amt;
+  };
+
+  undelegate = (val: Validator, amt: number) => {
+    if (this.delegation[val] < amt) {
+      this.m.events.push(Event.INSUFFICIENT_SHARES);
+      return;
+    }
+    this.tokens[val] -= amt;
+    this.delegation[val] -= amt;
+    const und: Undelegation = {
+      val: val,
+      creationHeight: this.m.h[P],
+      completionTime: this.m.t[P] + UNBONDING_SECONDS_P,
+      balance: amt,
+      initialBalance: amt,
+      onHold: true,
+      opID: this.opID,
+      expired: false,
```

Comment:
I don't like matured because it is not clear if that means the unbonding maturations have been received from every consumer. Expired means it has been processed by the staking module EndBLock. I will change the name as such.

Reply:
Updated code

https://github.com/cosmos/interchain-security/blob/0e3a9db8e7619f3d83f2bf8a711ee7c7d898258c/diff-tests/core/model/src/model.ts#L240-L248